File: android/app/src/main/kotlin/ca/etsmtl/applets/etsmobile/services/models/Session.kt
package ca.etsmtl.applets.etsmobile.services.models

import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.StringReader

val sessionPath = listOf("soap:Envelope", "soap:Body", "listeSessionsResponse", "listeSessionsResult", "liste", "Trimestre")
val sessionErrorPath = listOf("soap:Envelope", "soap:Body", "listeSessionsResponse", "listeSessionsResult", "erreur")

data class Session(
    val shortName: String?,
    val name: String?,
    val startDate: String?,
    val endDate: String?,
    val endDateCourses: String?,
    val startDateRegistration: String?,
    val deadlineRegistration: String?,
    val startDateCancellationWithRefund: String?,
    val deadlineCancellationWithRefund: String?,
    val deadlineCancellationWithRefundNewStudent: String?,
    val startDateCancellationWithoutRefundNewStudent: String?,
    val deadlineCancellationWithoutRefundNewStudent: String?,
    val deadlineCancellationASEQ: String?
) {
    companion object {
        fun fromXml(xml: String): Session {
            val factory = XmlPullParserFactory.newInstance()
            factory.isNamespaceAware = true
            val parser = factory.newPullParser()
            parser.setInput(StringReader(xml))

            var eventType = parser.eventType
            var currentTag: String? = null
            var shortName: String? = null
            var name: String? = null
            var startDate: String? = null
            var endDate: String? = null
            var endDateCourses: String? = null
            var startDateRegistration: String? = null
            var deadlineRegistration: String? = null
            var startDateCancellationWithRefund: String? = null
            var deadlineCancellationWithRefund: String? = null
            var deadlineCancellationWithRefundNewStudent: String? = null
            var startDateCancellationWithoutRefundNewStudent: String? = null
            var deadlineCancellationWithoutRefundNewStudent: String? = null
            var deadlineCancellationASEQ: String? = null

            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        currentTag = parser.name
                    }
                    XmlPullParser.TEXT -> {
                        when (currentTag) {
                            "abrege" -> shortName = parser.text
                            "auLong" -> name = parser.text
                            "dateDebut" -> startDate = parser.text
                            "dateFin" -> endDate = parser.text
                            "dateFinCours" -> endDateCourses = parser.text
                            "dateDebutChemiNot" -> startDateRegistration = parser.text
                            "dateFinChemiNot" -> deadlineRegistration = parser.text
                            "dateDebutAnnulationAvecRemboursement" -> startDateCancellationWithRefund = parser.text
                            "dateFinAnnulationAvecRemboursement" -> deadlineCancellationWithRefund = parser.text
                            "dateFinAnnulationAvecRemboursementNouveauxEtudiants" -> deadlineCancellationWithRefundNewStudent = parser.text
                            "dateDebutAnnulationSansRemboursementNouveauxEtudiants" -> startDateCancellationWithoutRefundNewStudent = parser.text
                            "dateFinAnnulationSansRemboursementNouveauxEtudiants" -> deadlineCancellationWithoutRefundNewStudent = parser.text
                            "dateLimitePourAnnulerASEQ" -> deadlineCancellationASEQ = parser.text
                        }
                    }
                    XmlPullParser.END_TAG -> {
                        currentTag = null
                    }
                }
                eventType = parser.next()
            }

            return Session(
                shortName,
                name,
                startDate,
                endDate,
                endDateCourses,
                startDateRegistration,
                deadlineRegistration,
                startDateCancellationWithRefund,
                deadlineCancellationWithRefund,
                deadlineCancellationWithRefundNewStudent,
                startDateCancellationWithoutRefundNewStudent,
                deadlineCancellationWithoutRefundNewStudent,
                deadlineCancellationASEQ
            )
        }
    }
}

File: lib/domain/models/signets-api/session.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';

part 'session.freezed.dart';
part 'session.g.dart';

@freezed
abstract class Session with _$Session {
  // ignore: unused_element to create methods
  const Session._();

  const factory Session({
    /// Short name (like H2020)
    @JsonKey(name: 'abrege', )
    required String shortName,

    /// Complete name (like Hiver 2020)
    @JsonKey(name: 'auLong')
    required String longName,

    /// Start date
    @JsonKey(name: 'dateDebut')
    required DateTime startDate,

    /// End date
    @JsonKey(name: 'dateFin')
    required DateTime endDate,

    /// Courses end date
    @JsonKey(name: 'dateFinCours')
    required DateTime endDateCourses,

    /// Registration start
    @JsonKey(name: 'dateDebutChemiNot')
    required DateTime startDateRegistration,

    /// Registration end
    @JsonKey(name: 'dateFinChemiNot')
    required DateTime deadlineRegistration,

    /// Cancellation with refund start
    @JsonKey(name: 'dateDebutAnnulationAvecRemboursement')
    required DateTime startDateCancellationWithRefund,

    /// Cancellation with refund end
    @JsonKey(name: 'dateFinAnnulationAvecRemboursement')
    required DateTime deadlineCancellationWithRefund,

    /// Cancellation with refund end (new students)
    @JsonKey(name: 'dateFinAnnulationAvecRemboursementNouveauxEtudiants')
    required DateTime deadlineCancellationWithRefundNewStudent,

    /// Cancellation without refund start (new students)
    @JsonKey(name: 'dateDebutAnnulationSansRemboursementNouveauxEtudiants')
    required DateTime startDateCancellationWithoutRefundNewStudent,

    /// Cancellation without refund end (new students)
    @JsonKey(name: 'dateFinAnnulationSansRemboursementNouveauxEtudiants')
    required DateTime deadlineCancellationWithoutRefundNewStudent,

    /// ASEQ cancellation deadline
    @JsonKey(name: 'dateLimitePourAnnulerASEQ')
    required DateTime deadlineCancellationASEQ,
  }) = _Session;

  factory Session.fromJson(Map<String, dynamic> json) => _$SessionFromJson(json);

  int get daysCompleted => DateTime.now().difference(startDate).inDays;
  int get totalDays => endDate.difference(startDate).inDays;
}
